src/frontend/Preprocessor.ml

```diff
@@ -5,16 +5,39 @@ open Lexing
 open Debugging
 module Str = Re.Str
 
-let dup_exists l =
-  match List.find_a_dup ~compare:String.compare l with
-  | Some _ -> true
-  | None -> false
-
 let include_stack = Stack.create ()
 let include_paths : string list ref = ref []
 let included_files : string list ref = ref []
 let size () = Stack.length include_stack
 
+let locations_map : (string * Middle.Location.t option) String.Table.t =
+  String.Table.create ()
+
+let reset_locations () = Hashtbl.clear locations_map
+
+let new_file_start_position filename included_from =
```

Comment:
The idea is that `new_file_start_position` is called once per included file; but that logic is fully contained in `Preprocessor`, so yes, the exported version could be combined with `reset_locations`. And I agree `Preprocessor.init` could be called earlier and be part of that as well.

At the start of `Parse.parse` the code initializes `Lexer.comments` and calls `Input_warnings.init`. Do you think those too should be handled by `Preprocessor.init`? (requires moving the relevant mutable state to `Preprocessor` in order to avoid dependency cycles--IMO reasonable for `Lexer.comments` at least)

And then there's the cleanup at the end of `Parse.parse`. The code clears out `Lexer.comments` after it's no longer needed but leaves the global state in `Input_warnings` in whatever state it was at the end of parsing. Neither of these really matter; if `parse` is called again it cleans up the state anyway, and if it's not called then the globals are never touched again so who cares.
Calling `reset_locations` there is pointless too, I now think.

Reply:
I think the `Input_warnings` state should stay put. I don't have much of an opinion on the comments - it's not really a part of "preprocessing" per se, but the fact that it's so closely tied to the `#include` """comments""" mean I can justify it.

I do think it's worth making the other clean-ups you mention